Build failure with GHC-8.6 and containers-0.6

~~~~
Building     enummapset-0.5.2.2
Failed to install enummapset-0.5.2.2
Build log ( ~/.cabal/logs/ghc-8.6.0.20180823/enummapset-0.5.2.2-2u0xauM3JVBCyceeGTpYwG.log ):
cabal: Entering directory '/tmp/cabal-tmp-26260/enummapset-0.5.2.2'
Configuring enummapset-0.5.2.2...
Preprocessing library for enummapset-0.5.2.2..
Building library for enummapset-0.5.2.2..
[1 of 5] Compiling Data.EnumSet     ( Data/EnumSet.hs, dist/build/Data/EnumSet.o )
[2 of 5] Compiling Data.EnumMap.Base ( Data/EnumMap/Base.hs, dist/build/Data/EnumMap/Base.o )
[3 of 5] Compiling Data.EnumMap.Strict ( Data/EnumMap/Strict.hs, dist/build/Data/EnumMap/Strict.o )
[4 of 5] Compiling Data.EnumMap.Lazy ( Data/EnumMap/Lazy.hs, dist/build/Data/EnumMap/Lazy.o )
[5 of 5] Compiling Data.EnumMap     ( Data/EnumMap.hs, dist/build/Data/EnumMap.o )

Data/EnumMap.hs:30:31: error:
    • Data.IntMap.insertWith' is gone. Use Data.IntMap.Strict.insertWith.
    • In the first argument of ‘(.)’, namely
        ‘I.insertWith' f (fromEnum k) x’
      In the second argument of ‘(.)’, namely
        ‘I.insertWith' f (fromEnum k) x . unWrap’
      In the expression:
        EnumMap . I.insertWith' f (fromEnum k) x . unWrap
   |
30 | insertWith' f k x = EnumMap . I.insertWith' f (fromEnum k) x . unWrap
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Data/EnumMap.hs:35:13: error:
    • Data.IntMap.insertWithKey' is gone. Use Data.IntMap.Strict.insertWithKey.
    • In the first argument of ‘(.)’, namely
        ‘I.insertWithKey' (f . toEnum) (fromEnum k) x’
      In the second argument of ‘(.)’, namely
        ‘I.insertWithKey' (f . toEnum) (fromEnum k) x . unWrap’
      In the expression:
        EnumMap . I.insertWithKey' (f . toEnum) (fromEnum k) x . unWrap
   |
35 |   EnumMap . I.insertWithKey' (f . toEnum) (fromEnum k) x . unWrap
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Data/EnumMap.hs:38:12: error:
    • Data.IntMap.fold' is gone. Use Data.IntMap.foldr or Prelude.foldr.
    • In the first argument of ‘(.)’, namely ‘I.fold f a’
      In the expression: I.fold f a . unWrap
      In an equation for ‘fold’: fold f a = I.fold f a . unWrap
   |
38 | fold f a = I.fold f a . unWrap
   |            ^^^^^^^^^^

Data/EnumMap.hs:41:19: error:
    • Data.IntMap.foldWithKey is gone. Use foldrWithKey.
    • In the first argument of ‘(.)’, namely
        ‘I.foldWithKey (f . toEnum) a’
      In the expression: I.foldWithKey (f . toEnum) a . unWrap
      In an equation for ‘foldWithKey’:
          foldWithKey f a = I.foldWithKey (f . toEnum) a . unWrap
   |
41 | foldWithKey f a = I.foldWithKey (f . toEnum) a . unWrap
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabal: Leaving directory '/tmp/cabal-tmp-26260/enummapset-0.5.2.2'
~~~~

--- a/Data/EnumMap.hs
+++ b/Data/EnumMap.hs
@@ -1,3 +1,7 @@
+-- Note that all these extensions are only needed for @Whoops@ & Co
+-- and only as long as "Utils.Containers.Internal.TypeError"
+-- is not exposed in the containers package
+-- (see https://github.com/haskell/containers/issues/586).
 {-# LANGUAGE CPP, DataKinds, FlexibleContexts, FlexibleInstances,
              KindSignatures, TypeFamilies, UndecidableInstances #-}
 -- |

--- a/Data/EnumMap.hs
+++ b/Data/EnumMap.hs
@@ -1,3 +1,5 @@
+{-# LANGUAGE CPP, DataKinds, FlexibleContexts, FlexibleInstances,
+             KindSignatures, TypeFamilies, UndecidableInstances #-}
 -- |
 -- Module      :  $Header$
 -- Description :  Data.IntMap with Enum keys.
@@ -13,29 +15,53 @@
 
 module Data.EnumMap
   ( module Data.EnumMap.Lazy
+#ifdef __GLASGOW_HASKELL__
+-- For GHC, we disable these, pending removal. For anything else,
+-- we just don't define them at all.
   , insertWith'
   , insertWithKey'
   , fold
   , foldWithKey
+#endif
   ) where
 
-import Prelude hiding ( filter, foldr, foldl, lookup, map, null )
-
 import Data.EnumMap.Lazy
-import Data.EnumMap.Base
 
-import qualified Data.IntMap as I
+#ifdef __GLASGOW_HASKELL__
+-- Unfortunately we can't access this module:
+-- import Utils.Containers.Internal.TypeError
+-- so we copy-paste things first:
 
-insertWith' :: (Enum k) => (a -> a -> a) -> k -> a -> EnumMap k a -> EnumMap k a
-insertWith' f k x = EnumMap . I.insertWith' f (fromEnum k) x . unWrap
+import GHC.TypeLits
 
-insertWithKey' :: (Enum k)
-  => (k -> a -> a -> a) -> k -> a -> EnumMap k a -> EnumMap k a
-insertWithKey' f k x =
-  EnumMap . I.insertWithKey' (f . toEnum) (fromEnum k) x . unWrap
+class Whoops (a :: Symbol)
 
-fold :: (a -> b -> b) -> b -> EnumMap k a -> b
-fold f a = I.fold f a . unWrap
+#if __GLASGOW_HASKELL__ >= 800
+instance TypeError ('Text a) => Whoops a
+#endif
 
-foldWithKey :: (Enum k) => (k -> a -> b -> b) -> b -> EnumMap k a -> b
-foldWithKey f a = I.foldWithKey (f . toEnum) a . unWrap
+
+-- | This function is being removed and is no longer usable.
+-- Use 'Data.EnumMap.Strict.insertWith'
+insertWith' :: Whoops "Data.EnumMap.insertWith' is gone. Use Data.EnumMap.Strict.insertWith."
+            => (a -> a -> a) -> k -> a -> EnumMap k a -> EnumMap k a
+insertWith' _ _ _ _ = undefined
+
+-- | This function is being removed and is no longer usable.
+-- Use 'Data.EnumMap.Strict.insertWithKey'.
+insertWithKey' :: Whoops "Data.EnumMap.insertWithKey' is gone. Use Data.EnumMap.Strict.insertWithKey."
+               => (k -> a -> a -> a) -> k -> a -> EnumMap k a -> EnumMap k a
+insertWithKey' _ _ _ _ = undefined
+
+-- | This function is being removed and is no longer usable.
+-- Use 'Data.EnumMap.Lazy.foldr'.
+fold :: Whoops "Data.EnumMap.fold' is gone. Use Data.EnumMap.foldr or Prelude.foldr."
+     => (a -> b -> b) -> b -> EnumMap k a -> b
+fold _ _ _ = undefined
+
+-- | This function is being removed and is no longer usable.
+-- Use 'foldrWithKey'.
+foldWithKey :: Whoops "Data.EnumMap.foldWithKey is gone. Use foldrWithKey."
+            => (k -> a -> b -> b) -> b -> EnumMap k a -> b
+foldWithKey _ _ _ = undefined
+#endif

--- a/enummapset.cabal
+++ b/enummapset.cabal
@@ -1,5 +1,5 @@
 name:           enummapset
-version:        0.5.2.2
+version:        0.6.0.1
 synopsis:       IntMap and IntSet with Enum keys/elements.
 description:    This package contains simple wrappers around 'Data.IntMap' and
                 'Data.IntSet' with 'Enum' keys and elements respectively.
@@ -37,7 +37,7 @@ Library
 
   build-depends:
     base < 5,
-    containers >= 0.5.2 && < 0.6,
+    containers >= 0.5.2 && < 0.7,
     semigroups >=0.1 && <1.0,
     deepseq
 

